Add grid sensors to SolarEdge_local

scheric opened the pull request:
## Description:
Added Grid voltage an grid frequency
Added string voltage
Added the amount of optimizers are live in the system

Added some attributions (inverters registered, inverter mode)

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ ] There is no commented out code in this PR.
  - [ ] I have followed the [development checklist][dev-checklist]


If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

scheric committed:
Add grid sensors

scheric commented:
@adamj12b you asked in https://github.com/home-assistant/home-assistant/issues/27065#issuecomment-536711545 to add some extra sensors. I do not have Import/Export meter sensors. 
Do you perhaps have these extra sensors and could you hand over the JSON so i can figure something out.

scheric commented:
#### Asking for advice
I noticed that there are no eccentricity constant aka V, A, Hz, 
Should i add those to the const.py

scheric committed:
Formatting

scheric committed:
Add possibility to add attributes

scheric committed:
Add optimizer attribute

scheric committed:
Remove bare 'except'

scheric committed:
add proper exception

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -167,6 +179,16 @@ def unit_of_measurement(self):
+                self._data.attr_name[self._json_key]: self._data.info[self._json_key]
+            }
+        except KeyError:
+            return {ATTR_ATTRIBUTION: 0}
```
Why would we return attribution 0?

scheric committed:
Remove return attribution 0

scheric committed:
Ad inverter attribution

scheric committed:
Style change

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -187,7 +187,7 @@ def device_state_attributes(self):
             }
         except KeyError:
-            return {ATTR_ATTRIBUTION: 0}
+            return
```
`return None`

scheric committed:
Add attribute name to sensors constants

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -40,71 +38,93 @@
         "Inverter Temperature",
         TEMP_CELSIUS,
         "mdi:thermometer",
+        "Operating mode",
```
State attribute keys should be lowercase snakecase.

scheric committed:
SENSOR_TYPES  alphabetical and snake_case lower

scheric committed:
Formatting

scheric commented:
Can the attributes overflow?
for example: By adding this all the individual temperatures will be in an attribute. If someone has like 200 optimizes will it break something?
```python
 self.info["optimizertemperature"] = temperature
```
Would someone actually be interested in that data?

---

> 
> #### Asking for advice
> 
> I noticed that there are no electricity constants aka [V], [A], [Hz],
> Should I add those to the `const.py` in a different PR of course.
> https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/const.py

scheric committed:
forgot snake_case lower

adamj12b commented:
@scheric Here is the info you're looking for. 

The meter info starts just after communication, there is 2 entries for metersList. The first one is the export meter, showing how much power is being pushed to the grid and the second one is import meter, power being pulled from the grid. Currently my system shows no power being imported so there is no currentPower entry for the import meter. totalEnergy for the metersList entries are how much power has passed in each direction. 

```
python3 ~/Documents/solaredge.py 
sn: "xxxxxxxxx"
site {
  productionWatt: 7679.09375
  sizeLimitWatt: 10000.0
  inverterStatus {
  }
}
powerWatt: 7679.09375
voltage: 250.94790649414062
frequencyHz: 60.02359390258789
optimizersStatus {
  total: 47
  online: 47
}
sOk: true
status: POWER_PRODUCTION
switchStatus: true
cosPhi: 1.0
powerLimit: 10000.0
country: 12
inverters {
  primary {
    dspSn: "xxxxxxxxx"
    voltage: 398.5516662597656
    optimizersStatus {
      total: 47
      online: 47
    }
    temperature {
      value: 115
      units {
        farenheit: true
      }
    }
    subSystem: 6144
  }
}
communication {
}
energy {
  today: 6200.744140625
  thisMonth: 218387.171875
  thisYear: 997130.3125
  total: 998177.0
}
portiaSubsystem: 768
metersList {
  currentPower: 6584.072265625
  totalEnergy: 313316.0
}
metersList {
  totalEnergy: 282992.0
}
serverComm {
  status {
    physicalConnection: true
    IP_Address: true
    Gateway_Link: true
    Internet_Link: true
    Monitoring_Link: true
    S_OK: true
  }
}
```

The entries im interested in seeing may have been covered already with these updates.
voltage
frequencyHz
optimizersStatus - total/online
status: POWER_PRODUCTION(changes with inverter mode)
and the export and import meter

![image](https://user-images.githubusercontent.com/3638834/66269823-41f75500-e81b-11e9-8a9a-edd367ac66c1.png)

scheric committed:
Add extra meter sensors

scheric committed:
add critical error for debugging

scheric commented:
@adamj12b 
Could you test this version?
I think it is sensible to make a backup and enable ssh or samba so you can revert if the system crashes.
If the system crashes or does not work take a look at the log in the fronted or in configuration folder.

scheric committed:
Update sensor.py

adamj12b commented:
@scheric No crashing issues.

It looks as though the export and import meters are swapped. It is showing as exporting power in HA but its actually importing that power.

Here is what I am seeing

![image](https://user-images.githubusercontent.com/3638834/66273010-0752e400-e83e-11e9-9d61-613464325290.png)
![image](https://user-images.githubusercontent.com/3638834/66273021-2baec080-e83e-11e9-88b0-9a4496c14d18.png)


And what the status json is showing.
```
parallels@debian-gnu-linux-vm:~$ python3 ~/Documents/solaredge.py 
sn: "xxxxxxxx"
site {
  productionWatt: 1660.0418701171875
  sizeLimitWatt: 10000.0
  inverterStatus {
  }
}
powerWatt: 1660.0418701171875
voltage: 244.09837341308594
frequencyHz: 60.02430725097656
optimizersStatus {
  total: 47
  online: 47
}
sOk: true
status: POWER_PRODUCTION
switchStatus: true
cosPhi: 1.0
powerLimit: 10000.0
country: 12
inverters {
  primary {
    dspSn: "xxxxxxxx"
    voltage: 388.71466064453125
    optimizersStatus {
      total: 47
      online: 47
    }
    temperature {
      value: 88
      units {
        farenheit: true
      }
    }
    subSystem: 6144
  }
}
communication {
}
energy {
  today: 19774.3203125
  thisMonth: 231960.75
  thisYear: 1010703.875
  total: 1011751.0
}
portiaSubsystem: 768
metersList {
  totalEnergy: 320848.0
}
metersList {
  currentPower: 594.0
  totalEnergy: 283486.0
}
serverComm {
  status {
    physicalConnection: true
    IP_Address: true
    Gateway_Link: true
    Internet_Link: true
    Monitoring_Link: true
    S_OK: true
  }
}

```
As you can see, the currentPower entry is flipping between what the system is doing. import/export. Currently the system is exporting because it is cloudy.

Additionally, the total import energy value is being displayed as total export energy and the total import energy entity does not exist.

adamj12b commented:
@scheric testing in the terminal, the import meter index is 1 and export is 0. The code has these using the opposite.

scheric committed:
swam meter sensors

scheric commented:
@adamj12b  
Thanks for testing! 👍 
I hope the values keep the same location during different request.

scheric committed:
Add suitable icons to meter reading

scheric committed:
Fix for pointless-statement

homeassistant/components/solaredge_local/sensor.py:173:8: W0104: Statement seems to have no effect (pointless-statement)
homeassistant/components/solaredge_local/sensor.py:192:8: W0104: Statement seems to have no effect (pointless-statement)
homeassistant/components/solaredge_local/sensor.py:349:16: W0104: Statement seems to have no effect (pointless-statement)
homeassistant/components/solaredge_local/sensor.py:356:16: W0104: Statement seems to have no effect (pointless-statement)

adamj12b commented:
@scheric I just realized that I am not getting an entity for Total Import Energy. Line 181 in sensor.py I believe may be the issue as the function appears to be names incorrectly. 

Thank you for all the work you have done to implement these new features!

scheric committed:
Rename import energy sensor

scheric commented:
@adamj12b
Could you test this version.

adamj12b commented:
@scheric Yes, the import sensor is now displaying and is correct.

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -122,8 +166,47 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+                "",
+            ]
+    except IndexError:
+        _LOGGER.critical("meter list 0 sensors not created")
```
Please use error level at the most.

MartinHjelmare replied:
We start logging messages with capital letter.

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -122,8 +166,47 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+                "",
+            ]
+    except IndexError:
+        _LOGGER.critical("meter list 1 sensors not created")
```
As above.

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -24,63 +24,107 @@
+        "Grid Frequency",
+        "Hz",
+        "mdi:current-ac",
+        "",
```
Use `None` to represent missing attribute.

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -167,6 +252,14 @@ def unit_of_measurement(self):
+        """Return the state attributes."""
+        try:
+            return {self._attr: self._data.info[self._json_key]}
+        except KeyError:
```
Is the KeyError and missing attribute key always correlated? Are we risking to return a dict with `None` as key?

scheric replied:
No i don't think so.
https://github.com/home-assistant/home-assistant/blob/58e5c553f2138a18d9d6b3f44040a9f98a94c5b5/homeassistant/components/solaredge_local/sensor.py#L346
In this line the data is created in the info dict with the name provided by: `self._json_key` in this case: `"optimizers"`
So if there is no data created in the info dict then it would return `KeyError` boucouse the name `self._json_key` does not exist.

MartinHjelmare replied:
Right, but are we sure that this `KeyError` case correlates to when `self._attr` is `None`? I'm worried since we use two different variables `self._attr` and self._json_key` but we only check one of them here with the `KeyError` exception, `self._json_key`.

scheric replied:
This is what the states menu looks like:
![states menu](https://user-images.githubusercontent.com/38077357/66413245-4f693680-e9f7-11e9-962a-bbf081d8468a.PNG)

scheric replied:
Does HA only check the first one? If that is so then it should not work.
but if I try this in an IDE
```python
import traceback

test1 = None
test2 = {}
test2["temperature"] = [
        "currentTemp",
        "Current temperature",
        'C',
        "mdi:oil-temperature"]

try:
    print("beginning try")
    print(test1)

    print(test2['temperature'])
    print(test2['nottemperature'])
    #traceback.print_exc()
except:
    pass
    traceback.print_exc()
```
```txt

beginning try
None
['currentTemp', 'Current temperature', 'C', 'mdi:oil-temperature']
Traceback (most recent call last):
  File "D:/Users/eric/OneDrive - Hanze/OneDrive - Hanzehogeschool Groningen/programeren/random test/append en add dicts test.py", line 108, in <module>
    print(test2['nottemperature'])
KeyError: 'nottemperature'

```

So if HA treats the `try` the same as my IDE it should work because printing `test1` that has value ` None` does not create a problem.

MartinHjelmare replied:
The `KeyError` happens if `self._json_key` is not in `self._data.info` dict. It doesn't take the `self._attr` value into account at all. We have to make that connection separately or add another check.

MartinHjelmare replied:
Currently the connection is done manually here:
```py
"optimizer_connected": [
        "optimizers",
        "Optimizers Online",
        "optimizers",
        "mdi:solar-panel",
        "optimizers_connected",
    ],
```
and here:
```py
self.info["optimizers"] = status.optimizersStatus.total
self.info["invertertemperature"] = INVERTER_MODES[status.status]
```
So if we would make an error in those two places and not match both places, we'll get a non expected result in the `device_state_attributes`. So if we can make this connection more strict, it would be less easy to make a mistake.

scheric replied:
Okay, I came to the conclusion that `self._attr` in `\/` does not create an error.   [picture below] https://github.com/home-assistant/home-assistant/blob/060ea848bc16173bddd8c76ee19bd5d8c5a94ca7/homeassistant/components/solaredge_local/sensor.py#L256-L262

It will always send Null to the attributes if a line like this exist.   [picture below]  https://github.com/home-assistant/home-assistant/blob/060ea848bc16173bddd8c76ee19bd5d8c5a94ca7/homeassistant/components/solaredge_local/sensor.py#L346-L347

So a problem occurs when a developer does not add an attribute name to the `SENSOR_TYPES` 
That error could be fixed with this:
```python
    @property
    def device_state_attributes(self):
        """Return the state attributes."""
        if self._attr:
            try:
                return {self._attr: self._data.info[self._json_key]}
            except KeyError:
                return None
```
With this code both places need to be occupied. So the sensor needs an attribute name and attribute data to be visible in the fronted.
Do you mean this check?


--------------------------
--------------------
--------------------


![null in sensor](https://user-images.githubusercontent.com/38077357/66422842-ba703880-ea0a-11e9-9551-f8dc7a9cd2ed.PNG)
This line was added: 
`self.info["optimizerpower"] = power`

scheric replied:
So a problem occurs when a developer does not add an attribute name to the SENSOR_TYPES
That error could be fixed with this:
```python
    @property
    def device_state_attributes(self):
        """Return the state attributes."""
        if self._attr:
            try:
                return {self._attr: self._data.info[self._json_key]}
            except KeyError:
                return None
```
With this code both places need to be occupied. So the sensor needs an attribute name and attribute data to be visible in the fronted.
**Do you mean this check?**
Correct me if I am wrong.

MartinHjelmare replied:
Yes, that's enough. We should `return None` in the other case.

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -243,6 +337,28 @@ def update(self):
+                    self.data["currentPowerimport"] = status.metersList[1].currentPower
+                    self.data["totalEnergyimport"] = status.metersList[1].totalEnergy
+            except IndexError:
+                _LOGGER.critical("no values send to fronted")
```
As above.

MartinHjelmare replied:
Please use correct grammar in the logging messages. Don't mention the frontend. The frontend is an optional representation of the state machine.

Write something like:
```py
"Missing values for meter blabla"
```

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -243,6 +337,28 @@ def update(self):
+                    self.data["currentPowerexport"] = status.metersList[0].currentPower
+                    self.data["totalEnergyexport"] = status.metersList[0].totalEnergy
+            except IndexError:
+                _LOGGER.critical("no values send to fronted")
```
As above.

scheric committed:
Insert feadback

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -195,17 +196,17 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
             ]
     except IndexError:
-        _LOGGER.critical("meter list 1 sensors not created")
+        _LOGGER.info("Export meter sensors are not created.")
```
If it's just a debug message, let's put it at debug level.

scheric committed:
Change to debug info

scheric committed:
Add check if attribute name exist

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -256,9 +256,12 @@ def unit_of_measurement(self):
+                return {self._attr: self._data.info[self._json_key]}
+            except KeyError:
+                return None
+        else:
```
We can remove the `else:` and outdent below.

scheric committed:
Remove unnecessary else

MartinHjelmare commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -167,6 +253,15 @@ def unit_of_measurement(self):
+                return {self._attr: self._data.info[self._json_key]}
+            except KeyError:
+                return None
+
```
We still need a return at the end for the other case.
```suggestion
        return None
```

scheric committed:
Add return None if no attributes

MartinHjelmare reviewed (approved):
Looks good!

scheric committed:
flake
